cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,68 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (field_arrays_.size() != other.fields().size()) { return false; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  if (null_count_ > 0) {
+    bool equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+    if (!equal_null_bitmap) { return false; }
+
+    bool fields_equal = true;
+    for (size_t i = 0; i < field_arrays_.size(); ++i) {
+      fields_equal = field_arrays_[i].get()->Equals(other.field(i));
+      if (!fields_equal) { return false; }
+    }
+  }
+  return true;
+}
+
+bool StructArray::Equals(const std::shared_ptr<Array>& arr) const {
+  if (this == arr.get()) { return true; }
+  if (!arr) { return false; }
+  if (this->type_enum() != arr->type_enum()) { return false; }
+  return EqualsExact(*static_cast<const StructArray*>(arr.get()));
+}
+
+Status StructArray::Validate() const {
+  if (length_ < 0) { return Status::Invalid("Length was negative"); }
+
+  if (null_count() > length_) {
+    return Status::Invalid("Null count exceeds the length of this struct");
+  }
+
+  if (field_arrays_.size() > 0) {
+    // Validate fields
+    int32_t array_length = field_arrays_[0]->length();
+    for (auto it : field_arrays_) {
+      if (!(it->length() == array_length)) {
+        std::stringstream ss;
+        ss << "Length is not equal from field " << it->type()->ToString();
+        return Status::Invalid(ss.str());
+      }
+
+      const Status child_valid = it->Validate();
+      if (!child_valid.ok()) {
+        std::stringstream ss;
+        ss << "Child array invalid: " << child_valid.ToString();
+        return Status::Invalid(ss.str());
+      }
+    }
+
+    // Validate null bitmap
+    // TODO: No checks for the consistency of bitmaps. Its cost maybe expensive.
```

Comment:
Is this actually a TODO? I think what we've been discussing is that there is no requirement for consistency of the child bitmaps (a null parent slot may "mask" a non-null child value).